### PR TITLE
refactor: rewrite `as_polars_df.data.frame`

### LINKED ISSUE #816: Rewrite `as_polars_df.data.frame`
Related to #676

### PR DESCRIPTION
**Pull request by eitsupi**
Close #816
It used to call `pl$DataFrame` internally, but has been completely rewritten to not call it.

Speeds seem to be comparable, but memory usage may have been reduced.
(The `bench` package does not measure memory usage on the Rust side, but given that the output RPolarsDataFrame is the same, memory usage on the Rust side should not be much different.)

``` r
big_df = do.call(rbind, lapply(1:10, \(x) nycflights13::flights))

library(polars)
library(nanoarrow)

bench::mark(
  pl_DataFrame = {
    pl$DataFrame(big_df)
  },
  as_polars_df = {
    as_polars_df(big_df)
  },
  from_nanoarrow = {
    as_polars_df(as_nanoarrow_array_stream(big_df))
  },
  check = FALSE,
  min_iterations = 10L
)
#> # A tibble: 3 × 6
#>   expression          min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>     <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 pl_DataFrame   826.41ms 839.49ms     1.16     1.91MB    0.289
#> 2 as_polars_df   738.85ms 864.31ms     1.16   204.05KB    0.129
#> 3 from_nanoarrow    1.15s    1.27s     0.790   77.48MB    0.339
```

<sup>Created on 2024-02-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by eitsupi** [comment]
I will update the test and documentation later.

@etiennebacher @sorhawell If you are interested, check the performance in the benchmarks above.

**Comment by etiennebacher** [comment]
I'll review in more details later, but I'm a bit confused about the objective of this PR. If the speed isn't affected and we don't know if the memory used decreases, then why rewrite this? Is this an intermediate step to fix #676?

**Comment by eitsupi** [comment]
Given the simplicity of the process, it is not surprising that memory usage is reduced.

File: NEWS.md [hunk 1]
```diff
@@ -8,6 +8,8 @@
     each column. The output of `$flags` for `Series` was also improved and now
     contains `FAST_EXPLODE` for `Series` of type `list` and `array` (#809).
 -   Add string methods `to_lowercase()` and `to_uppercase()` for `Series` (#810).
+-   `as_polars_df()` for `data.frame` is completely rewritten and new arguments
```
Comment (etiennebacher) [inline NEWS.md:11]:
```suggestion
-   `as_polars_df()` for `data.frame` is more memory-efficient and new arguments
```

**Review by etiennebacher — APPROVED** [review]

**Pull request merged**